### ISSUE TITLE
Tomcat: ClassNotFoundException when JARs with Valve are in WEB-INF/lib instead of tomcat/lib?

### ISSUE THREAD

**Issue by alshain** [dblock/waffle#371]
The Tomcat [guide](https://github.com/dblock/waffle/blob/master/Docs/tomcat/TomcatMixedSingleSignOnAndFormAuthenticatorValve.md) says the JARs can be placed _eithr_ into `WEB-INF/lib` or into `tomcat/lib`.

[This stackoverflow answer](http://stackoverflow.com/questions/6138396/custom-tomcat-valve-contained-in-web-app-war-file) claims that custom Valves don't work when placed in `WEB-INF/lib` because a different classloader is used, but from the linked [Tomcat documentation](http://tomcat.apache.org/tomcat-8.0-doc/class-loader-howto.html) (link updated for Tomcat 8), I cannot see anything that would explain it, except for the fact that different class loaders exist.

However, if I copy the JARs described in the article from `WEB-INF/lib` to `tomcat/lib`, everything works.

Is this a mistake in the guide, or are there other factors at play?

I'm using tomcat 8.0.36, JDK 8 x64

**Comment by dblock** [comment]
What do we do in the demos? If this is incorrect would appreciate if you could make a PR fixing the documentation.

**Comment by alshain** [comment]
Assuming you mean the demo referenced [here][demo], first, there is no folder Samples/waffle-negotioate, there is a war. I've made a fresh Tomcat instance without custom JARs and dropped `waffle-negotiate-1.8.0.war` into `webapps`.

[Upon boot, I got the same error message.][gist]

Looking at its `WEB-INF/context.xml`, it also uses the Valve as described in the guide.

```
<?xml version='1.0' encoding='utf-8'?>
<!DOCTYPE Context>
<Context>
    <Valve className="waffle.apache.NegotiateAuthenticator" principalFormat="fqn" roleFormat="both"
        protocols="Negotiate,NTLM" />
    <Realm className="waffle.apache.WindowsRealm" />
</Context>
```

[demo]: https://github.com/dblock/waffle/blob/master/Docs/tomcat/TomcatSingleSignOnValve.md#waffle-authenticator-demo

[gist]: https://gist.github.com/alshain/e0279ba06c8fc8549b51d78b3c8e898d

**Comment by dblock** [comment]
Been a while since I touched this. Trust you to dig this all up and clarify docs, thanks!

**Comment by alshain** [comment]
I can't find an authoritative source on this,  though. I'm a bit puzzled as to why the demos have been this way for such a presumably long time, as well as the documentation. 

Surely this had been tested before?

**Comment by dblock** [comment]
It could be that things changed with Tomcat 8, but also almost nobody uses valves, most people use a filter.

**Comment by alshain** [comment]
When would I want to use which? 

The first link in the stackoverflow.com post talks about tomcat 6. I'll test it.

**Comment by hazendaz** [comment]
If using valve jars must be in tomcat/lib.  If using filter the go into application/lib.  Last i tested both were working in tomcat 8.  Not sure when ill get some time to reconfirm.

Get Outlook for Android<https://aka.ms/ghei36>



On Tue, Jul 12, 2016 at 8:59 AM -0400, "Chris" <notifications@github.com<mailto:notifications@github.com>> wrote:


The Tomcat guide<https://github.com/dblock/waffle/blob/master/Docs/tomcat/TomcatMixedSingleSignOnAndFormAuthenticatorValve.md> says the JARs can be placed eithr into WEB-INF/lib or into tomcat/lib.

This stackoverflow answer<http://stackoverflow.com/questions/6138396/custom-tomcat-valve-contained-in-web-app-war-file> claims that it doesn't worked when placed in WEB-INF/lib because a different classloader is used, but from the linked Tomcat documentation<http://tomcat.apache.org/tomcat-8.0-doc/class-loader-howto.html> (link updated for Tomcat 8), I cannot see anything that would explain it, except for the fact that different class loaders exist.

However, if I copy the JARs described in the article from WEB-INF/lib to tomcat/lib, everything works.

Is this a mistake in the guide, or are there other factors at play?

I'm using tomcat 8.0.36, JDK 8 x64

-

**Comment by dblock** [comment]
Valves are Tomcat-specific, filters are some Java web standard thing. Your mileage may vary :)

**Comment by alshain** [comment]
@hazendaz I see, it's just that the guide says they can go into either directory when using Valves, and I was wondering whether I was doing something else wrong.

@dblock I also tried with Tomcat 7 and Tomcat 8.5, same result.

I noticed that I haven't looked closely enough at the demos. I noticed that waffle-tomcat.jar is missing from the demo WARs, which makes sense as it apparently needs to go directly into Tomcat's WAR directory.
